x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchMaintenanceService.java

```diff
@@ -54,16 +58,34 @@
     private final AtomicBoolean isClosed = new AtomicBoolean(false);
     private volatile Scheduler.Cancellable cancellable;
 
-    AsyncSearchMaintenanceService(String localNodeId,
+    AsyncSearchMaintenanceService(ClusterService clusterService,
+                                  String localNodeId,
                                   Settings nodeSettings,
                                   ThreadPool threadPool,
                                   AsyncSearchIndexService indexService) {
+        this.clusterService = clusterService;
         this.localNodeId = localNodeId;
         this.threadPool = threadPool;
         this.indexService = indexService;
         this.delay = ASYNC_SEARCH_CLEANUP_INTERVAL_SETTING.get(nodeSettings);
     }
 
+    @Override
+    protected void doStart() {
+        clusterService.addListener(this);
+    }
+
+    @Override
+    protected void doStop() {
+        stopInternal();
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+        stopInternal();
+        isClosed.compareAndSet(false, true);
```

Comment:
Just a suggestion, I'm unsure, but would it be possible to use the `lifecyle` state that comes with AbstractLifecycleComponent now instead of this internal `isClosed` flag? To me it looks AbstractLifecycleComponent#close() moves the state to CLOSED before calling `doClose()` here. Please ignore if I'm wrong, would like to understand the difference in that case.

Reply:
That's a good suggestion, I pushed another commit to apply the change that you're suggesting.